components/codec/src/columnar/mod.rs

```diff
@@ -0,0 +1,550 @@
+
+    #[snafu(display("Failed to do compact encoding, err:{source}"))]
+    CompactEncode { source: crate::compact::Error },
+}
+
+define_result!(Error);
+
+/// The trait bound on the encoders for different types.
+trait ValuesEncoder<T> {
+    /// Encode a batch of values into the `buf`.
+    ///
+    /// As the `estimated_encoded_size` method is provided, the `buf` should be
+    /// pre-allocate.
+    fn encode<B, I>(&self, buf: &mut B, values: I) -> Result<()>
+    where
+        B: BufMut,
+        I: Iterator<Item = T>;
+
+    /// The estimated size for memory pre-allocated.
+    fn estimated_encoded_size<I>(&self, values: I) -> usize
+    where
+        I: Iterator<Item = T>,
+    {
+        let (lower, higher) = values.size_hint();
+        let num = lower.max(higher.unwrap_or_default());
+        num * std::mem::size_of::<T>()
+    }
+}
+
+/// The trait bound on the decoders for different types.
+trait ValuesDecoder<T> {
+    fn decode<B, F>(&self, buf: &mut B, f: F) -> Result<()>
+    where
+        B: Buf,
+        F: FnMut(T) -> Result<()>;
+}
+
+/// The implementation for [`ValuesEncoder`].
+struct ValuesEncoderImpl;
+
+/// The implementation for [`ValuesDecoder`].
+struct ValuesDecoderImpl;
+
+#[derive(Clone, Debug)]
+pub struct ColumnarEncoder {
+    column_idx: u32,
+}
+
+/// A hint helps column encoding.
+pub struct EncodeHint {
+    pub num_nulls: Option<usize>,
+    pub num_datums: Option<usize>,
+    pub datum_kind: DatumKind,
+}
+
+impl EncodeHint {
+    fn compute_num_nulls<'a, I>(&mut self, datums: &I) -> usize
+    where
+        I: Iterator<Item = DatumView<'a>> + Clone,
+    {
+        if let Some(v) = self.num_nulls {
+            v
+        } else {
+            let num_nulls = datums.clone().filter(|v| v.is_null()).count();
+            self.num_nulls = Some(num_nulls);
+            num_nulls
+        }
+    }
+
+    fn compute_num_datums<'a, I>(&mut self, datums: &I) -> usize
+    where
+        I: Iterator<Item = DatumView<'a>> + Clone,
+    {
+        if let Some(v) = self.num_datums {
+            v
+        } else {
+            let num_datums = datums.clone().count();
+            self.num_datums = Some(num_datums);
+            num_datums
+        }
+    }
+}
+
+impl ColumnarEncoder {
+    const VERSION: u8 = 0;
+
+    pub fn new(column_idx: u32) -> Self {
+        Self { column_idx }
+    }
+
+    /// The header includes `version`, `datum_kind`, `column_idx`, `num_datums`
+    /// and `num_nulls`.
+    ///
+    /// Refer to the [encode](ColumnarEncoder::encode) method.
+    #[inline]
+    const fn header_size() -> usize {
+        1 + 1 + 4 + 4 + 4
+    }
+
+    /// The layout of the final serialized bytes:
+    /// ```plaintext
+    /// +-------------+----------------+-----------------+-----------------+----------------+---------------+---------------------+
+    /// | version(u8) | datum_kind(u8) | column_idx(u32) | num_datums(u32) | num_nulls(u32) | nulls_bit_set | non-null data block |
```

Comment:
How about put nulls_bit_set at the end of buf, then when decode, we  will combine decode_without_nulls and decode_with_nulls into one function.

Reply:
I guess it's good to separate the two methods according to whether null values are contained.